packages/cli/src/telemetry.ts

```diff
@@ -0,0 +1,146 @@
+/// <reference lib="dom" />
+import fs from 'fs-extra';
+import uuid from 'uuid/v4';
+import path from 'path';
+import crypto from 'crypto';
+import envPaths from 'env-paths';
+import mapValues from 'lodash.mapvalues';
+import pickBy from 'lodash.pickby';
+import inquirer from 'inquirer';
+import firebase from 'firebase/app';
+import 'firebase/auth';
+import 'firebase/firestore';
+import { DISABLE_INTERACTIVITY } from './prompts/prompt';
+
+import ProjectFile from './models/files/ProjectFile';
+
+const FIREBASE_CONFIG = {
+  apiKey: 'AIzaSyAZoGB2p26UejODezZPmczgwehI6xlSKPs',
+  authDomain: 'cli-telemetry.firebaseapp.com',
+  databaseURL: 'https://cli-telemetry.firebaseio.com',
+  projectId: 'cli-telemetry',
+  storageBucket: '',
+  messagingSenderId: '449985678611',
+  appId: '1:449985678611:web:88b411adc68e6521e19ee6',
+};
+
+type Field = string | number | boolean;
+
+interface GlobalTelemetryOptions {
+  optIn: boolean;
+  uuid: string;
+  salt: string;
+}
+
+export default {
+  async report(commandName: string, options: any, interactive: boolean): Promise<void> {
+    const telemetry = await checkOptIn(interactive);
+    if (telemetry === undefined || !telemetry.optIn) return;
+
+    // extract network name if present
+    let { network } = pickBy(options, (_, key) => key === 'network');
+    if (network) delete options.network;
+    if (network && network.match(/dev-/)) network = 'development';
+
+    // encrypt data before sending it
+    const concealedData = concealData(commandName, options, telemetry.salt);
+    const commandData = network ? { ...concealedData, network } : concealedData;
+
+    await this.sendToFirebase(telemetry.uuid, commandData);
+  },
+
+  async sendToFirebase(uuid: string, commandData: any): Promise<void> {
+    // Initialize Firebase and anonymously authenticate
+    const app = firebase.initializeApp(FIREBASE_CONFIG);
+    const db = app.firestore();
+    const { FieldValue } = firebase.firestore;
+
+    try {
+      await app.auth().signInAnonymously();
+
+      // create a new command document for the current uuid
+      await db.runTransaction(async tx => {
+        const dbSnapshot = await tx.get(db.doc(`users/${uuid}`));
+        let incrementalId;
+        // if the current user document exists, retreive the latest command id and create a new command document.
+        // otherwise, create a document for the user and set the id to 0.
+        if (dbSnapshot.exists) {
+          incrementalId = dbSnapshot.get('latestId') + 1;
+          await tx.update(db.doc(`users/${uuid}`), { latestId: FieldValue.increment(1) });
+        } else {
+          incrementalId = 0;
+          await tx.set(db.doc(`users/${uuid}`), { latestId: 0 });
+        }
+        await tx.set(db.collection(`users/${uuid}/commands`).doc(), { ...commandData, id: incrementalId });
+      });
+
+      // close all connections
+      await app.delete();
+    } catch (_) {
```

Comment:
I feel we do poor job tracking our exceptions. I suggest we start using service like sentry.io for collecting anonymous exceptions data. In that way we'll know errors happen even before user will report them as issue. Let's discuss it.

Reply:
I agree, we do need to manage exceptions in a better way. In this particular case, we decided to just `return` to avoid throwing errors caused by telemetry, which I believe is fine.